File: Programmers/[17679] 프렌즈4블록/woong-jae/17679.js
function solution(m, n, board) {
    let answer = 0;
    const dr = [-1, 0, 1, 0], dc = [0, 1, 0, -1];
    
    const isValid = (row, col) => {
        if(0 <= row && row < m && 0 <= col && col < n) return true;
        return false;
    }
    
    const mark = (row, col, marked) => {
        const target = board[row][col];
        for(let direction = 0; direction < 4; direction++) {
            let r1 = row + dr[direction], c1 = col + dc[direction];
            let r2 = row + dr[(direction + 1) % 4], c2 = col + dc[(direction + 1) % 4];
            let r3 = row + dr[direction] + dr[(direction + 1) % 4], c3 = col + dc[direction] + dc[(direction + 1) % 4];
            if(!isValid(r1, c1) || !isValid(r2, c2)) continue;
            if(target === board[r1][c1] && target === board[r2][c2] && target === board[r3][c3]) {
                marked[row][col] = marked[r1][c1] = marked[r2][c2] = marked[r3][c3] = true;
            }
        }
    }   
    
    board = board.map(row => row.split(""));
    while(1) {
        let count = 0;
    
        const marked = Array.from(Array(m), () => Array(n).fill(false));
        // 모든 보드를 순회하면서 2 * 2 블록을 찾아 마크해놓음
        for(let row = 0; row < m; row++) {
            for(let col = 0; col < n; col++) {
                if(!board[row][col]) continue;
                mark(row, col, marked);
            }
        }
        // 마크한 블록을 지우고 보드 업데이트
        for(let col = 0; col < n; col++) {
            for(let row = m - 1; row >= 0; row--) {
                if(marked[row][col]) {
                    count++;
                    board[row][col] = null;
                    continue;
                }
                let cur = row;
                while(isValid(cur + 1, col) && !board[cur + 1][col]) cur++;
                if(cur !== row) {
                    [board[row][col], board[cur][col]] = [board[cur][col], board[row][col]];
                }
            }
        }
        
        if(count === 0) break;
        answer += count;
    }
    
    return answer;
}

File: Programmers/[17679] 프렌즈4블록/JFe/17679.py
def solution(m, n, board):
    answer, temp, play = 0, [], True
    for i, b in enumerate(board):
        board[i] = list(b)
    # 더이상 지울게 없을 때까지 반복
    while play:
        play = False
        temp = []
        # 블록 확인
        for i in range(m-1):
            for j in range(n-1):
                # 2X2 형태면 temp에 append
                if board[i][j] != 0 and board[i][j] == board[i+1][j] == board[i][j+1] == board[i+1][j+1]:
                    temp.append((i, j))
                    play = True
        # 블록 지우기
        for i, j in temp:
            board[i][j] = 0
            board[i+1][j] = 0
            board[i][j+1] = 0
            board[i+1][j+1] = 0
        # 빈 공간 채우기
        for i in range(m-1):
            for j in range(n):
                # 밑에 빈 공간 있으면 내리기
                if board[i][j] != 0 and board[i+1][j] == 0:
                    board[i+1][j], board[i][j] = board[i][j], board[i+1][j]
                    # 해당 블록 위에 블록들도 내려주기
                    for k in range(i-1, 0, -1):
                        if board[k][j] != 0:
                            board[k][j], board[k+1][j] = board[k+1][j], board[k][j]
    # 지운 블록 수 체크
    for i in range(m):
        for j in range(n):
            if board[i][j] == 0:
                answer += 1
    return answer